app/src/main/java/fi/metatavu/muisti/exhibitionui/pages/components/MaterialTabLayoutComponentFactory.kt

```diff
@@ -0,0 +1,427 @@
+            )
+
+            initializeTabContentComponents(
+                buildContext = buildContext,
+                parent = parent,
+                data = data,
+                tabContentComponents = tabContentComponents
+            )
+        }
+
+    }
+
+    /**
+     * Adds listener for tab changes
+     *
+     * @param view view
+     * @param tabContentComponents tab content components
+     */
+    private fun addTabChangeListener(view: MuistiTabLayout, tabContentComponents: List<View>) {
+        view.addOnTabSelectedListener(object : TabLayout.OnTabSelectedListener {
+
+            override fun onTabReselected(tab: TabLayout.Tab?) {
+            }
+
+            override fun onTabUnselected(tab: TabLayout.Tab?) {
+            }
+
+            override fun onTabSelected(tab: TabLayout.Tab?) {
+                val position = tab?.position ?: 0
+                tabContentComponents.forEachIndexed { index, tabContentComponent ->
+                    run {
+                        if (index == position) {
+                            tabContentComponent.visibility = View.VISIBLE
+                        } else {
+                            tabContentComponent.visibility = View.GONE
+                        }
+                    }
+                }
+            }
+        })
+    }
+
+    /**
+     * Constructs tab content components
+     *
+     * @param buildContext build context
+     * @param view view
+     * @param data tabs data
+     * @return tab content components
+     */
+    private fun constructTabContentComponents(buildContext: ComponentBuildContext, view: MuistiTabLayout, data: TabData): List<View> {
+        val pageLayoutView = buildContext.pageLayoutView
+
+        return data.tabs.mapIndexed { index, tabData ->
+            run {
+                val tab = view.newTab()
+                tab.text = tabData.label
+                view.addTab(tab)
+
+                val tabContentComponent = ImageViewComponentFactory().buildComponent(
+                    ComponentBuildContext(
+                        context = buildContext.context,
+                        parents = buildContext.parents.plus(view),
+                        pageLayoutView = PageLayoutView(
+                            id = "${pageLayoutView.id}/$index",
+                            children = emptyArray(),
+                            properties = tabData.properties,
+                            name = "${pageLayoutView.name}/$index",
+                            widget = PageLayoutWidgetType.mediaView
+                        ),
+                        page = buildContext.page.copy(
+                            resources = tabData.resources
+                        ),
+                        lifecycleListeners = buildContext.lifecycleListeners,
+                        visitorSessionListeners = buildContext.visitorSessionListeners
+                    )
+                )
+
+                tabContentComponent
```

Comment:
Missing something?

Reply:
Nope. Thats how you return values for Kotlin Lambdas :)

Reply:
Roger